### PR TITLE
Accelerated shift

### PR DESCRIPTION
**Pull request by tchaton**
Hello Sir,

I have added the new code over modified_* files. We will clean everything when validated on your side.

Best,

### CODE REVIEW

**Comment by Zhaoyi-Yan** [comment]
Hi, tchaton,
It is a nice PR. However, you seem to add some irrelevant files in. Files in folder `.idea` should not be pushed. Also files in `.ipynb_checkpoints`  seem duplicate. Please clean the PR first, then I will review carefully.

File: util/NonparametricShift.py [hunk 1]
```diff
@@ -4,6 +4,52 @@
...
+        for i in range(input_window.size(0)):
+            input_window[i] = input_window[i]*(1/(input_window[i].norm(2)+1e-8))
+            return input_window
+
+    def _norm(self, input_window):
+        for i in range(input_window.size(0)):
+            input_window[i] = input_window[i]*(1/(input_window[i].norm(2)+1e-8))
+        return input_window        
+    
+    def _paste(self, img, patch_size, stride, flag, patch):
```
Comment (Zhaoyi-Yan) [inline util/NonparametricShift.py]: `_paste` and `_extract_patches_from_flag` have duplicated code, `Line 34~Line 42` with `Line 9~18`. Can they share code?

File: util/util.py [hunk 2]
```diff
@@ -9,6 +9,68 @@
 import collections
 import math
 import torch.nn as nn
+import torch.nn.functional as F
+
+
+class OptimizerMask:
```
Comment (Zhaoyi-Yan) [inline util/util.py]: This class seems not to be used anywhere. Wonder whether the adoption of this class has been forgotten somewhere else.

File: util/util.py [hunk 3]
```diff
@@ -9,6 +9,68 @@
...
+        #plt.imshow(masks.reshape((64, 64)))
+        area_coverage = np.sum(masks) / np.product(masks.shape)
+        # print(area_coverage)
+        if area_coverage < self.stop_criteria:
+            return False
+        else:
+            return True
+
+    def mean(self):
+        _mean = np.mean(np.sum(self.masks[self.indexes], axis=-1)) / (64 * 64)
```
Comment (Zhaoyi-Yan) [inline util/util.py]: `64*64` seems very hard here. How about random masks?

File: data/aligned_dataset_resized.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+#-*-coding:utf-8-*-
```
[the remark this answers is not in the document]
Comment (tchaton) [reply data/aligned_dataset_resized.py]: For my dataset, every image should be resized. It is to the user choice to define which dataset loader he wants. The code is supporting both options.

File: models/InnerShiftTripleFunction.py [hunk 5]
```diff
@@ -43,12 +43,16 @@ def forward(ctx, input, mask, shift_sz, stride, triple_w, flag, nonmask_point_id
...
-            latter = latter_all.narrow(0, idx, 1)
-            former = former_all.narrow(0, idx, 1)
+            latter = latter_all.narrow(0, idx, 1) ### UNET ADD
+            former = former_all.narrow(0, idx, 1) ### UPCONV
+
             Nonparm = NonparametricShift()
             _, conv_enc, conv_new_dec, _, = Nonparm.buildAutoencoder(latter.squeeze(), False, False, nonmask_point_idx, shift_sz, stride)
 
+            
+            print(former.shape)
```
[the remark this answers is not in the document]
Comment (tchaton) [reply models/InnerShiftTripleFunction.py]: Would do.

**Comment by tchaton** [comment]
I have cleaned a little bit more, and finished implementing the soft attention based on the previous work. However, it is working only when detaching the gradients.

**Comment by Zhaoyi-Yan** [comment]
Maybe you can add `*xml` to  .gitignore

**Comment by Zhaoyi-Yan** [comment]
I just merge this in and then pull another PR to modify some parts of it.

**Comment by Zhaoyi-Yan** [comment]
Hi, tchaton
SoftShiftTriple seems haven't been implemented totally. See: https://github.com/Zhaoyi-Yan/Shift-Net_pytorch/blob/master/models/models.py#L10-L18.

Please delete `__init__.py` in models folder, and rename 'models/models.py' to `models/__init__.py`, also change https://github.com/Zhaoyi-Yan/Shift-Net_pytorch/blob/master/train.py#L4 to `from models import create_model`.

After doing this, I found the accelerate_unet_shift_triple doesn't work for `center` mask. The original unet_shift_triple works fine however is slower than the initial version.
